core/unix/os.c

```diff
@@ -1099,6 +1106,45 @@ get_application_short_name(void)
     return get_application_name_helper(false, false /* short name */);
 }
 
+/* Sets pointers to the application's command-line arguments. These pointers are then used
+ * by get_app_args().
+ */
+void
+set_app_args(IN int *app_argc_in, IN char **app_argv_in)
+{
+    app_argc = app_argc_in;
+    app_argv = app_argv_in;
+}
+
+/* Returns the number of application's command-line arguments. */
+int
+num_app_args()
+{
+    return *app_argc;
+}
+
+/* Returns the application's command-line arguments. */
+int
+get_app_args(OUT dr_app_arg_t *args_buf, int buf_size)
+{
+    if (args_buf == NULL || buf_size < 0) {
+#ifdef CLIENT_INTERFACE
+        dcontext_t *dcontext = get_thread_private_dcontext();
+        dcontext->client_data->error_code = DR_ERROR_INVALID_PARAM;
+#endif
+        return -1;
+    }
+
+    int num_args = num_app_args();
+    int min = (buf_size < num_args) ? buf_size : num_args;
+    for (int i = 0; i < min; i++) {
+        args_buf[i].start = (void *)app_argv[i];
```

Comment:
This will crash for late injection?  Check up front and return the not-impl error code.

Reply:
Done.